File: ios/YamabikoChat/Features/Fusion/FusionTracePresentation.swift
import Foundation

enum FusionTracePresentation {
    static func shortModelLabel(_ model: String) -> String {
        let trimmed = model.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return trimmed }
        if let slash = trimmed.lastIndex(of: "/") {
            return String(trimmed[trimmed.index(after: slash)...])
        }
        if trimmed.count > 24 {
            return String(trimmed.prefix(22)) + "…"
        }
        return trimmed
    }

    static func confidenceLabel(_ confidence: FusionConfidence) -> String {
        switch confidence {
        case .low:
            return L10n.text("低信頼")
        case .medium:
            return L10n.text("中信頼")
        case .high:
            return L10n.text("高信頼")
        }
    }

    static func phaseLabel(_ phase: FusionPhase) -> String {
        switch phase {
        case .panel:
            return L10n.text("パネル")
        case .judge:
            return L10n.text("ジャッジ")
        case .synthesizer:
            return L10n.text("合成")
        case .fallback:
            return L10n.text("フォールバック")
        }
    }

    static func progressPhaseTitle(_ snapshot: FusionProgressSnapshot) -> String {
        switch snapshot.phase {
        case .panel:
            return L10n.format(
                "パネル実行中 (%d/%d)",
                snapshot.completedPanelCount,
                snapshot.totalPanelCount
            )
        case .judge:
            return L10n.text("ジャッジ中")
        case .synthesizer:
            return L10n.text("回答を合成中")
        case .fallback:
            return L10n.text("フォールバックで回答中")
        }
    }

    static func summaryLine(for trace: FusionTrace) -> String {
        let successCount = trace.panelResults.filter(\.success).count
        let totalPanels = trace.panelResults.count
        var parts = [
            L10n.format("Fusion · %d/%d panels", successCount, totalPanels)
        ]
        if let analysis = trace.judgeResult?.analysis {
            parts.append(confidenceLabel(analysis.confidence))
        }
        if let latencyMs = trace.totalLatencyMs {
            parts.append(formatLatency(ms: latencyMs))
        }
        if let cost = trace.totalCost {
            parts.append(String(format: "$%.4f", cost))
        }
        return parts.joined(separator: " · ")
    }

    static func formatLatency(ms: Int64) -> String {
        if ms < 1_000 {
            return L10n.format("%d ms", ms)
        }
        let seconds = Double(ms) / 1_000.0
        return L10n.format("%.1f s", seconds)
    }

    static func formatCost(_ cost: Double?) -> String? {
        guard let cost else { return nil }
        return String(format: "$%.4f", cost)
    }

    static func formatTokens(input: Int?, output: Int?) -> String? {
        guard input != nil || output != nil else { return nil }
        let inputValue = input ?? 0
        let outputValue = output ?? 0
        return L10n.format("%d in / %d out", inputValue, outputValue)
    }
}

File: app/src/main/java/com/porarri/yamabikochat/ui/fusion/FusionTracePresentation.kt
package com.porarri.yamabikochat.ui.fusion

import com.porarri.yamabikochat.data.fusion.FusionConfidence
import com.porarri.yamabikochat.data.fusion.FusionPhase
import com.porarri.yamabikochat.data.fusion.FusionProgressSnapshot
import com.porarri.yamabikochat.data.fusion.FusionTrace

object FusionTracePresentation {
    fun shortModelLabel(model: String): String {
        val trimmed = model.trim()
        if (trimmed.isEmpty()) return trimmed
        val slash = trimmed.lastIndexOf('/')
        if (slash >= 0 && slash < trimmed.lastIndex) {
            return trimmed.substring(slash + 1)
        }
        return if (trimmed.length > 24) trimmed.take(22) + "…" else trimmed
    }

    fun confidenceLabel(confidence: FusionConfidence): String = when (confidence) {
        FusionConfidence.low -> "低信頼"
        FusionConfidence.medium -> "中信頼"
        FusionConfidence.high -> "高信頼"
    }

    fun phaseLabel(phase: FusionPhase): String = when (phase) {
        FusionPhase.panel -> "パネル"
        FusionPhase.judge -> "ジャッジ"
        FusionPhase.synthesizer -> "合成"
        FusionPhase.fallback -> "フォールバック"
    }

    fun progressPhaseTitle(snapshot: FusionProgressSnapshot): String = when (snapshot.phase) {
        FusionPhase.panel ->
            "パネル実行中 (${snapshot.completedPanelCount}/${snapshot.totalPanelCount})"
        FusionPhase.judge -> "ジャッジ中"
        FusionPhase.synthesizer -> "回答を合成中"
        FusionPhase.fallback -> "フォールバックで回答中"
    }

    fun summaryLine(forTrace: FusionTrace): String {
        val successCount = forTrace.panelResults.count { it.success }
        val totalPanels = forTrace.panelResults.size
        val parts = mutableListOf("Fusion · $successCount/$totalPanels panels")
        forTrace.judgeResult?.analysis?.let {
            parts.add(confidenceLabel(it.confidence))
        }
        forTrace.totalLatencyMs?.let { parts.add(formatLatency(it)) }
        forTrace.totalCost?.let { parts.add(String.format("$%.4f", it)) }
        return parts.joinToString(" · ")
    }

    fun formatLatency(ms: Long): String {
        if (ms < 1_000) return "$ms ms"
        return String.format("%.1f s", ms / 1000.0)
    }

    fun formatCost(cost: Double?): String? =
        cost?.let { String.format("$%.4f", it) }
}
